sel/constant.hpp

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include "sel/contract.hpp"
+#include "sel/strongly_ordered.hpp"
+#include "sel/term.hpp"
+
+#include <compare>
+#include <concepts>
+#include <format>
+#include <type_traits>
+
+namespace sel {
+
+/// symbol that represents a specified quantity in an expression
+template <class T>
+  requires std::signed_integral<T> or std::floating_point<T>
+class constant
+{
+  T value_{};
+
+public:
+  constexpr explicit constant(T value)
+      : value_{[&] {
+          if constexpr (not strongly_ordered<T>) {
+            SEL_PRECONDITION(value == value);
+          }
+          return std::move(value);
+        }()}
+  {}
+
+  template <class Self>
+  [[nodiscard]]
+  constexpr auto value(this Self&& self) -> auto&&
+  {
+    return std::forward<Self>(self).value_;
+  }
+
+  [[nodiscard]]
+  friend auto
+  operator==(const constant& x, const constant& y) -> bool = default;
+
+  template <class U>
+    requires std::equality_comparable_with<T, U>
+  [[nodiscard]]
+  friend constexpr auto
+  operator==(const constant& x, const constant<U>& y) -> bool
+  {
+    return x.value() == y.value();
+  }
+
+  [[nodiscard]]
+  friend constexpr auto
+  operator<=>(const constant& x, const constant& y) -> std::strong_ordering
+  {
+    const auto cmp = x.value() <=> y.value();
+    if constexpr (strongly_ordered<T>) {
+      return cmp;
+    } else {
+      if (cmp < 0) {
+        return std::strong_ordering::less;
+      }
+      if (cmp > 0) {
+        return std::strong_ordering::greater;
+      }
+      return std::strong_ordering::equal;
+    }
+  }
```

Comment:
can we reflect-based derive this sort of thing yet?

Reply:
It looks like no compilers implement reflection yet.

For `int` as the `value_type`, we can default it:

```cpp
[[nodiscard]]
friend constexpr auto
operator<=>(const constant& x, const constant&) requires strongly_ordered<T> = default;
```

However, this won't work for `float` since three-way comparison returns a `partial_ordering`.

Here we ignore the possibility that `cmp` is `std::partial_ordering::unordered` and use the same implementation for both `int` and `float`.